platform/posix/transport/src/openssl_posix.c

```diff
@@ -729,55 +721,96 @@ int32_t Openssl_Recv( NetworkContext_t * pNetworkContext,
 {
     OpensslParams_t * pOpensslParams = NULL;
     int32_t bytesReceived = 0;
-    int32_t sslError = 0;
 
     if( ( pNetworkContext == NULL ) || ( pNetworkContext->pParams == NULL ) )
     {
         LogError( ( "Parameter check failed: pNetworkContext is NULL." ) );
     }
-    else if( pNetworkContext->pParams->pSsl != NULL )
+    else if( pNetworkContext->pParams->pSsl == NULL )
+    {
+        LogError( ( "Failed to receive data over network: "
+                    "SSL object in network context is NULL." ) );
+    }
+    else
     {
+        int32_t pollStatus = 1, readStatus = 1, sslError = 0;
+        uint8_t shouldRead = 0U;
+        struct pollfd pollFds;
         pOpensslParams = pNetworkContext->pParams;
-        /* SSL read of data. */
-        bytesReceived = ( int32_t ) SSL_read( pOpensslParams->pSsl,
-                                              pBuffer,
-                                              ( int32_t ) bytesToRecv );
+
+        /* Initialize the file descriptor. */
+        pollFds.events = POLLIN | POLLPRI | POLLHUP | POLLERR;
+        pollFds.revents = 0;
+        /* Set the file descriptor for poll. */
+        pollFds.fd = pOpensslParams->socketDescriptor;
+
+        if( ( bytesToRecv > 1 ) || ( SSL_pending( pOpensslParams->pSsl ) > 0 ) )
+        {
+            shouldRead = 1U;
+        }
+        else
+        {
+            /* Non-speculative read so check if there is data to read from the socket. */
+            pollStatus = poll( &pollFds, 1, 0 );
+        }
+
+        if( pollStatus < 0 )
+        {
+            bytesReceived = -1;
+        }
+        else if( pollStatus == 0 )
+        {
+            /* No data available to be read from the socket. */
+            bytesReceived = 0;
+        }
+        else
```

Comment:
There should probably a check that the `pollFds.revent` output parameter has been set to `POLLIN` based on the following documentation for positive return value from `poll` in the [Linux man page](https://man7.org/linux/man-pages/man2/poll.2.html):
"_On success, poll() returns a nonnegative value which is the number of elements in the pollfds whose revents fields have been set to a nonzero value (indicating an event or an error).  A return value of zero indicates that the system call timed out before any file descriptors became read._"

Reply:
Also, an appropriate log message can be added if the `POLLERR` flag is set in the output parameter of `revents`

Reply:
We had a chat that the error in such cases would already be handled by `recv` for POSIX and `SSL_read` for OpenSSL, so it is not necessary to check these flags.